cpp/src/arrow/util/gather_internal.h

```diff
@@ -0,0 +1,287 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include "arrow/array/data.h"
+#include "arrow/util/bit_block_counter.h"
+#include "arrow/util/bit_run_reader.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/macros.h"
+
+// Implementation helpers for kernels that need to load/gather fixed-width
+// data from multiple, arbitrary indices.
+//
+// https://en.wikipedia.org/wiki/Gather/scatter_(vector_addressing)
+
+namespace arrow::internal {
+inline namespace gather_internal {
```

Comment:
This inner namespace doesn't need to be referenced. I never have to write `arrow::internal::gather_internal::Gather` — it's `arrow::internal::Gather`.

This was necessary to avoid performance regressions. I actually benchmarked this and it made a difference. My guess is that without the inline namespace, the SROA [1] optimization wasn't kicking in — the member variables of this class should be scalars placed in registers accessed from the loop.

Not only the `gather.Execute` call is inlined, the construction of `Gather` lowers to simply configuring registers.

Many STL classes live in inline namespaces like this so they can fully disappear at the code that uses them.

[1] [Scalar Replacement of Aggregates](https://llvm.org/docs/Passes.html#sroa-scalar-replacement-of-aggregates)

Reply:
> Many STL classes live in inline namespaces like this so they can fully disappear at the code that uses them.

I guess I don't understand why using an inline namespace would open up more optimization opportunities. Do you have a reference that explains this? Or is it just a side effect of whatever compiler version you're using?

Reply:
> I guess I don't understand why using an inline namespace would open up more optimization opportunities.

You understand that everything inside an anonymous namespace `namespace { ... }` has internal linkage. All the inline namespace symbols in a header when included in a `.cc` becomes internal to that compilation unit.

The compiler doesn't have to expose internal linkage symbols, so the decision for inlining becomes easier: if it's called only once, it's binary-size economical to inline it. The compiler might not inline it if multiple calls are made in that compilation unit.

Reply:
The reference does not seem to mention internal linkage. Here is how it describes [inline namespaces](https://en.cppreference.com/w/cpp/language/namespace#Inline_namespaces):
> Members of an inline namespace are treated as if they are members of the enclosing namespace in many situations (listed below).

Reply:
Also, I do not find any reference online supporting the idea that moving functions into inline namespaces can improve performance by changing linkage.

Reply:
I'm removing the `inline namespace` as it seems to be unnecessary in the presence of the force-inline on the `Execute*()` functions.